ppci/format/elf/file.py

```diff
@@ -79,26 +202,64 @@ def load(f):
         elf_file.read_strtab(f)
         for section in elf_file.sections:
             section.read_data(f)
-            section.name = elf_file.get_str(section.header["sh_name"])
+            section.name = elf_file.get_str(
+                section.header["sh_name"], elf_file.strtab
+            )
+            typ = SectionHeaderType(section.header["sh_type"])
+            if typ == SectionHeaderType.REL:
+                f.seek(section.header["sh_offset"])
+                rh = elf_file.header_types.RelocationTableEntry.read(f)
+                elf_file.relocations.append(ElfRelocation(rh, bits=bits))
+            elif typ == SectionHeaderType.RELA:
+                f.seek(section.header.sh_offset)
+                rh = elf_file.header_types.RelocationTableEntryWA.read(f)
+                elf_file.relocations.append(ElfRelocation(rh, bits=bits))
+            elif typ in [SectionHeaderType.SYMTAB, SectionHeaderType.DYNSYM]:
+                elf_file.symbole_table = elf_file.read_symtab(section)
+            elif typ == SectionHeaderType.STRTAB and section.name in [
+                ".strtab",
+                ".dynstr",
+            ]:
+                elf_file.symstrtab = section.read_data(f)
+
+        if "symbole_table" in vars(elf_file):
+            for relocation in elf_file.relocations:
+                relocation.connect_section(
+                    elf_file.symbole_table, elf_file.sections
+                )
+            for symbol in elf_file.symbole_table:
+                symbol.name = elf_file.get_str(
+                    symbol.header["st_name"], elf_file.symstrtab
+                )
+                if symbol.header.st_shndx < len(elf_file.sections):
+                    symbol.section = elf_file.sections[symbol.header.st_shndx]
         return elf_file
 
     def read_strtab(self, f):
         self.strtab = self.sections[self.elf_header.e_shstrndx].read_data(f)
 
-    def read_symbol_table(self, sym_section):
+    def read_symtab(self, sym_section):
         f = io.BytesIO(sym_section.data)
         count = (
             len(sym_section.data) // self.header_types.SymbolTableEntry.size
         )
         table = [
-            self.header_types.SymbolTableEntry.read(f) for _ in range(count)
+            ElfSymbol(self.header_types.SymbolTableEntry.read(f), i)
+            for i in range(count)
         ]
         return table
 
-    def get_str(self, offset):
+    def read_symbol_table(self, f):
+        return [s.header for s in self.symbole_table]
+
+    def get_str(self, offset, *strtab):
```

Comment:
Hmm, I see the purpose of passing the strtab along here. Maybe we could introduce a `StringTable` class, which has the get_str method? Then we can modify `read_strtab` such that it is cached, using functools.cache decorate. `read_strtab` should then return a new `StringTable` class.